strongbox-web-core/src/main/java/org/carlspring/strongbox/services/DirectoryListingServiceImpl.java

```diff
@@ -69,59 +79,80 @@ public DirectoryListing fromRepositories(Map<String, ? extends Repository> repos
+                                                                            (String) null);
+
+        List<Path> contentPaths = fetchContent(repositoryPath);
+
+        return generateDirectoryListing(null, contentPaths, directoryLinkBuilder, fileLinkBuilder);
     }
 
-    private DirectoryListing fromPath(Path path)
-        throws IOException
+    @Override
+    public DirectoryListing fromPath(UriComponentsBuilder directoryLinkBuilder,
+                                     UriComponentsBuilder fileLinkBuilder,
+                                     Path rootPath,
+                                     Path requestedPath)
+            throws IOException, RuntimeException
     {
-        path = path.normalize();
+        rootPath = rootPath.normalize();
+        requestedPath = requestedPath.normalize();
 
-        DirectoryListing directoryListing = new DirectoryListing();
+        if (!requestedPath.equals(rootPath) && !requestedPath.startsWith(rootPath))
+        {
+            String message = "Requested directory listing is outside the scope of the root path! Possible intrusion attack or misconfiguration!";
+            logger.error(message);
+            throw new RuntimeException(message);
+        }
 
-        Map<String, List<FileContent>> content = generateDirectoryListing(path);
+        Path relativePath = rootPath.relativize(requestedPath);
 
-        directoryListing.setDirectories(content.get("directories"));
-        directoryListing.setFiles(content.get("files"));
+        List<Path> contentPaths = fetchContent(requestedPath);
 
-        return directoryListing;
+        return generateDirectoryListing(relativePath, contentPaths, directoryLinkBuilder, fileLinkBuilder);
     }
 
-    private Map<String, List<FileContent>> generateDirectoryListing(Path path)
-        throws IOException
+    /**
+     * Generate a directory listing for a path.
+     *
+     * @param relativePath         A relative path to the artifact within the scope of (directory|file)LinkBuilder
+     *                             This field is only needed/used when listing non-storage paths.
+     * @param contentPaths         List of Paths to iterate over to generate the DirectoryListing.
+     * @param directoryLinkBuilder UriComponentsBuilder to be used for generating directory links
+     * @param fileLinkBuilder      UriComponentsBuilder to be used for generating download links;
+     *                             If null - file.url will be null as well.
+     *
+     * @return
+     *
+     * @throws IOException
+     */
+    private DirectoryListing generateDirectoryListing(Path relativePath,
+                                                      List<Path> contentPaths,
+                                                      UriComponentsBuilder directoryLinkBuilder,
+                                                      UriComponentsBuilder fileLinkBuilder)
+            throws IOException
```

Comment:
This method has been partially rewritten so it can work with a `UriCopmonentsBuilder` and is a bit more flexible when generating links for directories and files. This is especially necessary since `browse` and `download` links are completely different (i.e. `browse - /api/browse/some/path`; `download - /storages/some/path/artifact.jar`)